exec/cnex/global.c

```diff
@@ -139,14 +227,57 @@ void strb(TExecutor *exec)
     push(exec->stack, cell_fromCString(v ? "TRUE" : "FALSE"));
 }
 
+void substring(struct tagTExecutor *exec)
+{
+    // ToDo: Implement UTF8
+    Number len = top(exec->stack)->number; pop(exec->stack);
+    Number start = top(exec->stack)->number; pop(exec->stack);
+    TString *ss   = top(exec->stack)->string;
+
+    assert(number_is_integer(start));
+    assert(number_is_integer(len));
+
+    size_t offset = number_to_uint32(start);
+    if (offset > ss->length) {
+        exec->rtl_raise(exec, "ArrayIndexException", "offset", BID_ZERO);
+        return;
+    }
+
+    size_t length = number_to_uint32(len);
+    if (length + offset > ss->length) {
+        exec->rtl_raise(exec, "ArrayIndexException", "length", BID_ZERO);
+        return;
+    }
+
+    TString *r = string_subString(ss, offset, length);
+    pop(exec->stack);
+    push(exec->stack, cell_fromString(r));
+    string_freeString(r);
+}
+
+void odd(struct tagTExecutor *exec)
+{
+    Number n = top(exec->stack)->number; pop(exec->stack);
+
+    if (!number_is_integer(n)) {
+        exec->rtl_raise(exec, "ValueRangeException", "odd() requires integer", BID_ZERO);
+        return;
+    }
+
+    push(exec->stack, cell_fromBoolean(number_is_odd(n)));
+}
+
 void ord(TExecutor *exec)
 {
     Cell *s = top(exec->stack);
 
     if (s->string->length != 1) {
+        pop(exec->stack);
```

Comment:
As discussed, it is not necessary to clean up the stack in this case. It's harmless to do this, but the pop will be taken care of automatically by the exception handler.

Reply:
Removed unnecessary pop.